3/n inter batch parallelism

tchaton opened the pull request:
## What does this PR do?

<!--
Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change.

If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

The following links the related issue to the PR (https://docs.github.com/en/free-pro-team@latest/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
-->

Blocked by https://github.com/PyTorchLightning/pytorch-lightning/pull/9047/

This PR adds remove entirely InterBatchProcessor and rely on DataFetcher. 

fixes #8867
fixes #8316

### Does your PR introduce any breaking changes? If yes, please list them.

<!-- FILL IN or None -->

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

carmocca commented on pytorch_lightning/loops/batch/training_batch_loop.py:
```diff
@@ -545,12 +550,16 @@ def _build_kwargs(self, batch: Any, batch_idx: int, opt_idx: int, hiddens: Optio
         """
         # enable not needing to add opt_idx to training_step
-        step_kwargs = OrderedDict([("batch", batch), ("batch_idx", batch_idx)])
+        step_kwargs = OrderedDict({"batch": batch})
```
Using list of tuples is important as dicts do not guarantee order in 3.6

tchaton replied:
Done !

carmocca commented on pytorch_lightning/loops/epoch/training_epoch_loop.py:
```diff
@@ -118,29 +112,17 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
-                batch_output = self.batch_loop.run(batch, self.batch_idx)
-
-            self.batch_progress.increment_processed()
+        batch_idx, (batch, is_last) = next(dataloader_iter)
+
+        # ------------------------------------
+        # TRAINING_STEP + TRAINING_STEP_END
+        # ------------------------------------
+        self.batch_progress.increment_ready()
+
+        with self.trainer.profiler.profile("run_training_batch"):
```
The comment is misplaced

```suggestion
        batch_idx, (batch, is_last) = next(dataloader_iter)
        
        self.batch_progress.increment_ready()

        # ------------------------------------
        # TRAINING_STEP + TRAINING_STEP_END
        # ------------------------------------
        with self.trainer.profiler.profile("run_training_batch"):
```

tchaton replied:
Done !

tchaton commented:
Hey @yifuwang @ananthsub, mind reviewing this PR.

I removed the need for returning `is_last`, added support for accumulate_grad_batches and automatic optimization, optionally passing the `batch_idx`.

Best,
T.C

awaelchli commented on pytorch_lightning/trainer/connectors/data_connector.py:
```diff
@@ -250,6 +245,40 @@ def detach_data(model: "pl.LightningModule") -> None:
+                attr.teardown()
+                delattr(self, attr_name)
+
+    def _check_training_step_requires_dataloader_iter(self) -> bool:
```
don't we want to add this to ConfigValidator?

tchaton replied:
Done !

awaelchli commented on tests/utilities/test_fetching.py:
```diff
@@ -176,6 +180,14 @@ def test_dataloader(self):
+        def __init__(self, data_fetcher_cls: Type):
+            self.data_fetcher_cls = data_fetcher_cls
+
+        def on_batch_start(self, trainer, *_) -> None:
+            assert isinstance(trainer.data_connector.train_data_fetcher, self.data_fetcher_cls)
```
note we are relying on the callback implementation for testing here. this is fragile

yifuwang commented:
Thanks @tchaton!

> I removed the need for returning is_last, added support for accumulate_grad_batches and automatic optimization, optionally passing the batch_idx.

IIUC when `dataloader_iter` is passed to `training_step`, the contract is:
- `training_step` can prefetch/preprocess arbitrary number of batches from `dataloader_iter`
- However, `training_step` should use and only use the `batch_idx`th batch to optimize the model

Is my understanding correct?

ananthsub commented on pytorch_lightning/loops/batch/training_batch_loop.py:
```diff
@@ -104,6 +104,11 @@ def run(self, batch: Any, batch_idx: int) -> AttributeDict:
         self.batch_outputs = None  # free memory
         return output
 
+    def on_stop_iteration(self):
+        # if a stop iteration is being triggered within the ``training_step``,
+        # forward the expection to the parent loop.
+        raise StopIteration
```
+1 - not clear why we need special handling for stop iteration

tchaton replied:
Here is the following situation.

If a StopIteration is being triggered within the training_step.

On master, this exception is being catched by the TrainingBatchLoop run function which break the while loop.
However, the TrainingEpochLoop isn't stopped and the training continues even if it should have stopped.

Let me refactor this within the on_run_end.

tchaton replied:
I updated based on @awaelchli request, but I prefer the previous solution.

ananthsub reviewed (commented):
in addition to @yifuwang 's question, should the `training_step` with `dataloader_iter` be responsible for handling the stop iteration, rather than the loop catching this & doing post-processing before `on_run_end` ? before the contract was that the training step would return the last_batch indicator for the batch processor to handle

tchaton commented on pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:
```diff
@@ -199,7 +199,11 @@ def update_eval_epoch_metrics(self) -> _EVALUATE_OUTPUT:
     def on_train_split_start(self, batch_idx: int, split_idx: int, split_batch: Any) -> None:
-        self.trainer._results.extract_batch_size(split_batch)
+        # when the user request `dataloader_iter`, we can't track the batch_size
+        # and this is left to user responsability.
```
Done !

tchaton commented on pytorch_lightning/loops/epoch/training_epoch_loop.py:
```diff
@@ -118,33 +112,22 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
+
+        self.batch_progress.increment_ready()
+
+        # ------------------------------------
+        # TRAINING_STEP + TRAINING_STEP_END
+        # ------------------------------------
```
Removed !

tchaton commented on pytorch_lightning/trainer/connectors/data_connector.py:
```diff
@@ -250,6 +245,40 @@ def detach_data(model: "pl.LightningModule") -> None:
                 loader.unpatch(model)
 
+    def teardown(self) -> None:
+        items = list(vars(self).items())
+        for attr_name, attr in items:
+            if isinstance(attr, AbstractDataFetcher):
+                attr.teardown()
+                delattr(self, attr_name)
```
Done !

tchaton commented:
> Thanks @tchaton!
> 
> > I removed the need for returning is_last, added support for accumulate_grad_batches and automatic optimization, optionally passing the batch_idx.
> 
> IIUC when `dataloader_iter` is passed to `training_step`, the contract is:
> 
> * `training_step` can prefetch/preprocess arbitrary number of batches from `dataloader_iter`
> * However, `training_step` should use and only use the `batch_idx`th batch to optimize the model
> 
> Is my understanding correct?

Dear @yifuwang, yes the `training_step` can prefetch/preprocess arbitrary number of batches from `dataloader_iter`.
Internally, Lightning tracks how many batches have been fetched out within the `training_step` and can optionally provide this value to the user through the `batch_idx`. 
This is more for consistency with current API.

justusschock commented on pytorch_lightning/loops/epoch/training_epoch_loop.py:
```diff
@@ -81,7 +75,7 @@ def done(self) -> bool:
     def connect(
         self,
-        batch_loop: BATCH_LOOP_TYPE = None,
+        batch_loop: TrainingBatchLoop = None,
```
```suggestion
        batch_loop: Optional[TrainingBatchLoop] = None,
```

tchaton replied:
Done !

awaelchli commented on pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:
```diff
@@ -70,6 +70,8 @@ def on_run_start(
         self._dl_max_batches = dl_max_batches
         self._num_dataloaders = num_dataloaders
 
+        self.dataloader_iter = enumerate(dataloader_iter, self.batch_progress.current.ready)
```
we don't have the attribute set in init

tchaton replied:
Done !

carmocca replied:
Isn't this redundant with `_prepare_dataloader_iter` in the epoch loop?

remove this in favor of the other?

tchaton replied:
Updated !

awaelchli commented on tests/core/test_metric_result_integration.py:
```diff
@@ -411,6 +411,7 @@ def training_step(self, batch, batch_idx):
             # However, below we will simulate a failure on `batch_idx=3`.
 
             if self.trainer.fit_loop.restarting:
+
```
```suggestion
```

tchaton replied:
Done !

tchaton commented on pytorch_lightning/loops/batch/training_batch_loop.py:
```diff
@@ -550,9 +550,13 @@ def _build_kwargs(self, batch: Any, batch_idx: int, opt_idx: int, hiddens: Optio
         step_kwargs = OrderedDict([("batch", batch), ("batch_idx", batch_idx)])
 
         lightning_module = self.trainer.lightning_module
+        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+
+        if is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True):
+            if not is_param_in_hook_signature(training_step_fx, "batch_idx"):
+                del step_kwargs["batch_idx"]
```
Trying out ! I believed the test were failing with this :)

tchaton commented on pytorch_lightning/core/lightning.py:
```diff
@@ -454,6 +454,11 @@ def log(
 
+        if is_param_in_hook_signature(self.training_step, "dataloader_iter", explicit=True) and batch_size is None:
+            raise MisconfigurationException(
+                "When the `dataloader_iter` is requested within the `training_step`, `batch_size` should be provided."
```
Done !

tchaton commented on pytorch_lightning/core/lightning.py:
```diff
@@ -454,6 +454,11 @@ def log(
                     f" of {list(self._metric_attributes.values())}"
                 )
 
+        if is_param_in_hook_signature(self.training_step, "dataloader_iter", explicit=True) and batch_size is None:
```
`dataloader_iter` is currently enabled only for training.

carmocca replied:
Still, this should be `if self.trainer.training and is_param_in ....`

tchaton commented on pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:
```diff
@@ -70,6 +71,8 @@ def on_run_start(
         self._dl_max_batches = dl_max_batches
```
Done !

carmocca commented on tests/checkpointing/test_trainer_checkpoint.py:
```diff
@@ -75,9 +75,6 @@ def validation_step(self, batch, batch_idx):
         results.append(deepcopy(trainer.callback_metrics))
         best_model_paths.append(trainer.checkpoint_callback.best_model_path)
 
-    for idx in range(len(results) - 1):
-        assert results[idx]["val_loss"] > results[idx + 1]["val_loss"]
-
```
Why is this removed? Does it not work anymore?

tchaton replied:
No it doesn't.

carmocca commented on tests/utilities/test_fetching.py:
```diff
@@ -189,19 +192,19 @@ def test_trainer_num_prefetch_batches(tmpdir):
         t0 = time()
         trainer = Trainer(**trainer_kwargs)
         trainer.fit(model)
+        assert isinstance(trainer.data_connector.train_data_fetcher, InterBatchParallelDataFetcher)
```
Why move the assertions into the timings?

tchaton replied:
Moved out !

carmocca commented on tests/utilities/test_fetching.py:
```diff
@@ -236,10 +242,152 @@ def training_step(self, dataloader_iter, batch_idx):
+        def training_step(self, dataloader_iter: Iterator, *args) -> STEP_OUTPUT:
+            output = super().training_step(dataloader_iter)
+            if self.trigger_stop_iteration and args[0] == EXPECT_NUM_BATCHES_PROCESSED:
+                raise StopIteration
```
@tchaton @awaelchli Is using a `StopIteration` here international or was it just meant as a `DummyException`?

This has the problem that it's the exception checked by the loops to stop, so after it is raised, the loops will try to run `on_run_end` but any further state done at the end of `advance` or `on_advance_end` will not have happened so things can be out of sync.

Blocking #9349

tchaton replied:
Yes, this is intentional, but it should work as expected.
